implement structured proposal generation using Handlebars + Gemini

ibarrasb opened the pull request:
## Description

This PR implements  **structured proposal generation flow (WAVE-69)**.  

- Builds the AI prompt using **Handlebars**, based on the recipe’s inputs and output definitions.  
- Uses the recipe’s `modelType` when calling Gemini.  
- Generates a JSON schema using the recipe’s `outputsWithWordCount` field names.  
- Calls Gemini for structured JSON and applies each field’s **word/character limits**.  
- Returns a fully assembled in-memory `GrantProposal` ready to display.

This update also adds proper error handling when the recipe is missing an ID or outputs, and includes unit tests to verify schema behavior, prompt generation, and trimming logic.

**Reviewers should focus on:**
- Prompt construction (`grantRecipeService.generatePromptWithInputs`)
- AI schema + limit enforcement (`grantProposalService.generate`)
- The updated AI wrapper in `grantAiService`
- Unit tests ensuring correct behavior and preventing regressions

---

## What type of PR is this? (check all applicable)

- [ ] Refactor  
- [x] Feature  
- [ ] Bug Fix  
- [ ] Optimization  
- [ ] Documentation Update  

---

## Related Tickets & Documents

- **WAVE-69 — Generate Structured Output Using AI**  
  https://digital-aid-seattle.atlassian.net/browse/WAVE-69

---

ibarrasb committed:
implement structured proposal generation using Handlebars + Gemini

jnakaso commented on src/services/grantProposalService.ts:
```diff
@@ -43,32 +60,67 @@ class GrantProposalService extends FirestoreService<GrantProposal> {
+    }
+
+    // Keep raw AI output for debugging or review
+    const textResponse = JSON.stringify(aiResult, null, 2);
```
There is no need to save a text response for structured output.

jnakaso reviewed (approved):
Need to change when the token string is created.

I was going to suggest a name change:
- prompt -> template
- tokenString -> prompt  
more closely aligns with how they are used.

jnakaso commented on src/services/grantProposalService.ts:
```diff
@@ -43,32 +60,67 @@ class GrantProposalService extends FirestoreService<GrantProposal> {
+    
+    for (const field of outputs) {
+      const raw = aiResult[field.name] ?? "";
+      structuredResponse[field.name] = applyOutputLimit(raw, field);
```
The output limits need to be included with  the prompt.  We want the AI  to self-limit the output, so that the wording is adjusted, not simply truncated.

jnakaso replied:
Rather than modifying the output, the application should provide counts to the user.

ibarrasb committed:
requested changes

jnakaso reviewed (changes requested):
couple of minor nits

jnakaso commented on src/services/grantProposalService.ts:
```diff
@@ -43,32 +60,67 @@ class GrantProposalService extends FirestoreService<GrantProposal> {
+    }
+
+    // Build the prompt using recipe data
+    const prompt = grantRecipeService.generatePromptWithInputs(recipe);
```
Prompts need to be created and saved separately from output generation.  This will allow the user to assess token counts before submitting.

ibarrasb committed:
updated changes

ibarrasb committed:
firebase build

ibarrasb commented:
This PR also cleans up a few leftover references that were causing the Firebase build to fail after the proposal model changes.

Removed uses of textResponse and tokenString, which are no longer part of the proposal/recipe types.

Updated the list page and tests to read from structuredResponse instead.

Aligned the recipe flow so the Handlebars template is compiled into a prompt once and then reused during proposal generation.

Nothing functionally new here. Just making sure the code, types, and build are all in sync

jnakaso commented on src/services/grantRecipeService.ts:
```diff
@@ -75,18 +90,38 @@ class GrantRecipeService extends FirestoreService<GrantRecipe> {
+    .join("\n");
+
+  return `
+${basePrompt}
```
The template should contain this, for example

```
Create a grant proposal including the following information:
 {{#each inputs}}{{key}} = {{value}},
{{/each}} where{{#each outputs}}{{#unless @first}} and{{/unless}} the {{name}} is constrained by a maximum {{maxSymbols}} of {{unit}} {{/each}}.
```

jnakaso replied:
We'll update the details page to include the constructed prompt

jnakaso reviewed (changes requested):
Need to:

1. remove output truncation.
2. prompt construction should use only the template.

ibarrasb committed:
Align proposal generation with template based prompt flow

ibarrasb committed:
Merge branch 'dev' into feature/ai-structured-output

ibarrasb committed:
firebase build

ibarrasb commented:
@jnakaso I addressed the requested changes by removing output truncation in GrantProposalService and relying on the recipe’s Handlebars template to enforce output limits instead.

Prompt construction now uses the template only, and prompts are generated/saved separately from proposal generation so token counts can be assessed before submitting.

While resolving the merge conflict with dev, I kept the existing updatePrompt flow to avoid breaking the recipe detail page and tests. That method is currently preserved only to keep the build passing and UI functional.

jnakaso reviewed (approved):
thx
